fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/layers/interning/ScopedInterningLayer.java

```diff
@@ -46,16 +44,13 @@
     private static final int STATE_SUBSPACE_KEY_SUFFIX = -10;
 
     @Nonnull
-    private final StringInterningLayer interningLayer;
+    private Function<FDBRecordContext, CompletableFuture<Subspace>> baseSubspaceGenerator;
     @Nonnull
-    private final Subspace baseSubspace;
+    private Function<FDBRecordContext, CompletableFuture<Subspace>> nodeSubspaceGenerator;
     @Nonnull
-    private final Subspace nodeSubspace;
+    private Function<FDBRecordContext, CompletableFuture<Subspace>> stateSubspaceGenerator;
```

Comment:
This generator approach is going to add a pretty significant amount of overhead.  With the most recent `KeySpace` API changes, calls like `toTupleAsync()` and `toSubspaceAsync()` will completely re-resolve the full tuple or subspace on each call -- that is, there is no longer any caching of the result beyond that which is provided by, say, the directory layer caching.  With these generators, you are going to be reconstructing these values on each call and, in addition, reconstructing the interning layer itself on every call, which is not what I think you intend.

Maybe you could do something similar, though, with something like:

    @Nullable
    private CompletableFuture<Subspace> stateSubspaceFuture;
    ...
    private CompletableFuture<Subspace> getStateSubspace(@Nonnull FDBRecordContext context) {
        if (stateSubspaceFuture == null) {
            stateSubspaceFuture = getNodeSubspace(context).thenApply(node -> new StringInterningLayer(node, true));
        }
        return stateSubspaceFuture;
    }

This way, once the future is complete, its value is no longer recomputed.

Reply:
Hmm yeah this is a good point. Maybe we should use `AsyncLoadingCache` (with no expiration) for the subspace values? The approach you're suggesting works except it suffers from the same potential flaw that old versions of the loading cache did: if the `stateSubspaceFuture` completes exceptionally it needs to be refetched, or worse if it hangs forever we need a mechanism to cancel and refetch the value. The loading cache should handle both of those, what do you think?

Reply:
Addressed this by adding a new `LazyAsyncValue` class, similar to `AsyncLoadingCache`